pimd/pim_bsm.c

```diff
@@ -0,0 +1,129 @@
+/*
+ * pim_bsm.c: PIM BSM handling routines
+ *
+ * Copyright (C) 2018-19 Vmware, Inc.
+ * Saravanan K
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; see the file COPYING; if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
+ * MA 02110-1301 USA
+ */
+#include "command.h"
+#include "vty.h"
+#include "if.h"
+#include "pimd.h"
+#include "pim_iface.h"
+#include "pim_vty.h"
+#include "pim_cmd.h"
+#include "pim_instance.h"
+#include "pim_rpf.h"
+#include "pim_hello.h"
+#include "pim_pim.h"
+#include "pim_nht.h"
+#include "pim_bsm.h"
+#include "pim_time.h"
+
+/* Functions forward declaration */
+static void pim_bs_timer_start(struct bsm_scope *scope, int bs_timeout);
+static int on_bs_timer(struct thread *t);
+static void pim_bs_timer_stop(struct bsm_scope *scope);
+
+static void pim_free_bsgrp_data(struct bsgrp_node * bsgrp_node)
+{
+	if (bsgrp_node->bsrp_list)
+		list_delete(&bsgrp_node->bsrp_list);
+	if (bsgrp_node->partial_bsrp_list)
+		list_delete(&bsgrp_node->partial_bsrp_list);
+
+	XFREE(MTYPE_PIM_BSGRP_NODE, bsgrp_node);
+}
+
+
+
+static void pim_bsm_node_free(struct bsm_info *bsm)
+{
+	if (bsm->bsm)
+		XFREE(MTYPE_PIM_BSM_PKT_VAR_MEM, bsm->bsm);
+	XFREE(MTYPE_PIM_BSM_INFO, bsm);
+}
+
+void pim_bsm_proc_init (struct pim_instance *pim)
+{
+
+	memset(&pim->global_scope, 0, sizeof(struct bsm_scope));
+
+	pim->global_scope.sz_id             = PIM_GBL_SZ_ID;
+	pim->global_scope.bsrp_table        = route_table_init();
+	pim->global_scope.accept_nofwd_bsm  = TRUE;
+	pim->global_scope.state             = NO_INFO;
+	pim->global_scope.pim               = pim;
```

Comment:
I don't see the pim pointer being used.  And since this scope is part of the pim data structure why do we need another pointer to it?

Reply:
Haven't uploaded all the code yet. just in the process of uploading commits. This is a back pointer. scope is the timer argument for bootstrap timer. when timer expired we need the pim pointer to be used for deleting the tracked nexthop. More commits to follow